MbsAPI/fLmd.c

```diff
@@ -61,1186 +56,1266 @@ int clock_gettime(int clockid, struct timespec* tp)
-                          (pLmdControl->pMbsFileHeader->iUsedWords)*2+sizeof(sMbsFileHeader));
-  pLmdControl->iBytes+=iReturn;
-
-  if(iUseOffset == LMD__INDEX) { fLmdOffsetResize(pLmdControl,iReturn/4); } // create and set first value
-  printf("fLmdPutOpen: %s. Bytes:%d over:%d table:%d large:%d.\n",
-         Filename,iBytes,iOver,iUseOffset,iLargeFile);
-  return(LMD__SUCCESS);
+    int32_t iReturn;
+    struct timespec clock;
+
+    memset(pLmdControl, 0, sizeof(sLmdControl));
+
+    // allocate header or take extern
+    if (pBuffHead == LMD__STANDARD_HEADER) {
+        pLmdControl->pMbsFileHeader = (sMbsFileHeader*)malloc(sizeof(sMbsFileHeader));
+        memset(pLmdControl->pMbsFileHeader, 0, sizeof(sMbsFileHeader));
+        pLmdControl->iInternHeader = 1;
+    } else {
+        pLmdControl->pMbsFileHeader = pBuffHead;
+        pLmdControl->iInternHeader = 0;
+    }
+
+    clock_gettime(CLOCK_REALTIME, &clock);
+    pLmdControl->pMbsFileHeader->iTimeSpecSec = clock.tv_sec;
+    pLmdControl->pMbsFileHeader->iTimeSpecNanoSec = clock.tv_nsec;
+
+    pLmdControl->pMbsFileHeader->iType = LMD__TYPE_FILE_HEADER_101_1;
+    pLmdControl->pMbsFileHeader->iEndian = 1;
+    size_t len = strlen(Filename);
+    if (len < sizeof(pLmdControl->cFile)) {
+        strncpy(pLmdControl->cFile, Filename, len);
+    } else {
+        strncpy(pLmdControl->cFile, Filename, sizeof(pLmdControl->cFile) - 1);
+    }
+    // optionally allocate buffer
+    if (iBytes > 0) {
+        pLmdControl->pBuffer = (int16_t*)malloc(iBytes);
+        pLmdControl->iInternBuffer = 1;
+    }
+    pLmdControl->iBufferWords = iBytes / 2;
+    pLmdControl->iLeftWords = iBytes / 2;
+    // open file
+    if (iOver == LMD__NO_OVERWRITE) {   // do not overwrite
+        if ((pLmdControl->fFile = (FILE*)fopen64(Filename, "r")) != NULL) {
+            printf("fLmdPutOpen: File exists: %s\n", Filename);
+            fLmdCleanup(pLmdControl);
+            fclose(pLmdControl->fFile);
+            return (PUTLMD__FILE_EXIST);
+        }
+    }
+
+    if ((pLmdControl->fFile = (FILE*)fopen64(Filename, "w+")) == NULL) {
+        printf("fLmdPutOpen: Error open file %s\n", Filename);
+        fLmdCleanup(pLmdControl);
+        return (PUTLMD__OPEN_ERR);
+    }
+
+    if (iLargeFile == LMD__LARGE_FILE) {
+        pLmdControl->iOffsetSize = 8;
+    } else {
+        pLmdControl->iOffsetSize = 4;
+    }
+    pLmdControl->pMbsFileHeader->iOffsetSize = pLmdControl->iOffsetSize;
+
+    // write header
+    iReturn = fLmdWriteBuffer(pLmdControl,
+                              (char*)pLmdControl->pMbsFileHeader,
+                              (pLmdControl->pMbsFileHeader->iUsedWords) * 2 + sizeof(sMbsFileHeader));
+    pLmdControl->iBytes += iReturn;
+
+    if (iUseOffset == LMD__INDEX) {
+        fLmdOffsetResize(pLmdControl, iReturn / 4);
+    }   // create and set first value
+    printf("fLmdPutOpen: %s. Bytes:%d over:%d table:%d large:%d.\n", Filename, iBytes, iOver, iUseOffset, iLargeFile);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [%d in format string (no. 5) requires 'int' but the argument type is 'unsigned int'.](https://app.codacy.com/manual/FairRootGroup/FairRoot/pullRequest?prid=5323596)